tests/5.1/target/test_target_has_device_addr.c

```diff
@@ -0,0 +1,53 @@
+//===--- test_target_has_device_addr.c --------------------------------------===//
+//
+// OpenMP API Version 5.1 Nov 2020
+//
+// This test verifies the 'has_device_addr' feature added to the target construct.
+// We tested this by checking a scalar & array address in one target region, and
+// verifying that the address remains the same in a second target region which
+// uses the 'has_device_addr' feature.
+//
+////===----------------------------------------------------------------------===//
+#include <assert.h>
+#include <omp.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include "ompvv.h"
+
+#define N 1000
+
+int test_target_has_device_addr() {
+  OMPVV_INFOMSG("test_target_has_device_addr");
+  int errors = 0;
+  int x = 10;
+  int arr[N];
+  for(int i=0;i<N;i++){
+      arr[i] = i;
+  }
+  int *first_scalar_device_addr;
+  int *first_arr_device_addr;
+  int *second_scalar_device_addr;
+  int *second_arr_device_addr;
+  #pragma omp target map(from: first_scalar_device_addr, first_arr_device_addr)
+  {
+    first_scalar_device_addr = &x;
+    first_arr_device_addr = &arr[0];
+  }
+  #pragma omp target map(from:second_scalar_device_addr, second_arr_device_addr) has_device_addr(x, arr[N]) 
+  {
+    second_scalar_device_addr = &x;
+    second_arr_device_addr = &arr[0];
+  }
```

Comment:
_Side note: GCC 12 will have has_device_addr support (C/C++/Fortran) – to be committed today._

Regarding the testcase: I think `arr` works with host fallback or (unified) shared memory. But `x` will likely fail reliably. Namely:

```C
  int x = 10; int arr[N];
```
There is nothing which puts `x` and `arr` into device memory. Thus, they only live on the host until ...
```C
   #pragma omp target map(from: first_scalar_device_addr, first_arr_device_addr)
    { first_scalar_device_addr = &x; first_arr_device_addr = &arr[0]; }
```
Here, the two variables are implicitly mapped – as `firstprivate(x) map(tofrom: arr)`. – Especially the `firstprivate(x)` will ensure that in the next target section, `x` is almost guaranteed to have a different address as here. As `arr` is previously not mapped, it also will be unmapped at the end of this target region.

And in the next target region, `x` and `arr` will keep their host value (due to the has_device_addr) ... 
```C
   #pragma omp target map(from:second_scalar_device_addr, second_arr_device_addr) has_device_addr(x, arr[N]) 
```
... – thus, '&x' and '&arr' should give the same value as on the host – but not a device addr. _(Ignoring issues with the semantic of 'device address' and that it will work with shared memory/host fallback.)_

How about something like the following. Unfortunately, the suggestions diff does not seem to show the changed lines. Thus, here's a list:
* Added: omp target enter/exit data
* Explicit mapping of x and arr for the first target region (actually, only required for 'x')
* Wrapped second region into a omp target data + use_device_addr
* Removed a `[N]` – cf. note below.
* Added data_x/data_arr → additional test.
```suggestion
  #pragma omp target enter data map(to: x, arr)
  #pragma omp target map(from: first_scalar_device_addr, first_arr_device_addr) map(to: x, arr)
  {
    first_scalar_device_addr = &x;
    first_arr_device_addr = &arr[0];
  }
  int *data_x, *data_arr;
  #pragma omp target data use_device_addr(x,arr)
  {
  data_x = &x; data_arr = &arr[0];
  #pragma omp target map(from:second_scalar_device_addr, second_arr_device_addr) has_device_addr(x, arr) 
  {
    second_scalar_device_addr = &x;
    second_arr_device_addr = &arr[0];
  }
  }
 #pragma omp target exit data map(release: x, arr)
 OMPVV_TEST_AND_SET(errors, first_scalar_device_addr != data_x);
 OMPVV_TEST_AND_SET(errors, first_arr_device_addr != data_arr);
```
**NOTES:**
* On the first omp target, instead of `map(to: x)`, a `defaultmap` could have been used instead. And instead of `to` using `present` would be also possible.
* I also removed the `[N]` from `has_device_addr(arr[N])`: while array sections are permitted, array elements aren't. _As an array section (for an has_device_addr list item) gives the same result as the full variable, the array section is only syntactical sugar. — Similarly for `use_device_addr`, except that that for the latter, the array section must enclose mapped array elements; thus, only if the whole array is mapped, the array section can be left out._

**Disclaimer:** I have not compiled neither the original version nor the version using my patch. And I might have missed some fine print.

Reply:
Hey @tob2 these changes look good! Just to clarify, could you explain why you decided to add in the target data use_device_addr? Would the code not function as expected without it?

Reply:
Well, in order that you can write `has_device_addr(x, arr)`, the two variables have to have a device address. The easiest way to achieve this is to use `target data ... use_device_addr(x, arr)` that both variables have the device address inside the target-data region.

In principle, something like `int *y = omp_target_alloc(...);` would also work, which then could be directly passed to `is_device_ptr()`, except that `y` is not a device address but only points to a device address (hence: is_device_ptr and not has_device_addr). I currently do not see how I can get rid of the 'target data', but probably using a reference type can achieve this somehow.

In contrast, for Fortran, something like the the following should work:
```F90
use iso_c_binding; use omp_lib
type(c_ptr) :: cptr
integer, pointer :: x(:,:)
cptr = omp_target_alloc (...)
call c_f_pointer(cptr, x, [10, 10])
!$omp target has_device_addr(x)
  x(:,:) = 42
!$omp end target
```